### ISSUE TITLE
unexpected error "Member must have length greater than or equal to 1"

### ISSUE THREAD

**Issue by nunofernandes** [fluent-plugins-nursery/fluent-plugin-cloudwatch-logs#116]
Hello,

I'm using:
```
2018-06-25 13:47:31 +0000 [info]: gem 'fluent-plugin-cloudwatch-logs' version '0.5.0'
2018-06-25 13:47:31 +0000 [info]: gem 'fluent-plugin-elasticsearch' version '2.10.1'
2018-06-25 13:47:31 +0000 [info]: gem 'fluent-plugin-elasticsearch' version '2.4.0'
2018-06-25 13:47:31 +0000 [info]: gem 'fluent-plugin-grok-parser' version '2.1.6'
2018-06-25 13:47:31 +0000 [info]: gem 'fluent-plugin-kafka' version '0.6.5'
2018-06-25 13:47:31 +0000 [info]: gem 'fluent-plugin-rewrite-tag-filter' version '2.0.1'
2018-06-25 13:47:31 +0000 [info]: gem 'fluent-plugin-s3' version '1.1.0'
2018-06-25 13:47:31 +0000 [info]: gem 'fluent-plugin-td' version '1.0.0'
2018-06-25 13:47:31 +0000 [info]: gem 'fluent-plugin-td-monitoring' version '0.2.3'
2018-06-25 13:47:31 +0000 [info]: gem 'fluent-plugin-webhdfs' version '1.2.2'
2018-06-25 13:47:31 +0000 [info]: gem 'fluentd' version '1.0.2'
```

And my fluentd is dying with:

```
2018-06-25 13:47:32 +0000 [warn]: #0 thread exited by unexpected error plugin=Fluent::Plugin::CloudwatchLogsInput title=:in_cloudwatch_logs_runner error_class=Aws::CloudWatchLogs::Errors::InvalidParameterException error="1 validation error detected: Value '' at 'nextToken' failed to satisfy constraint: Member must have length greater than or equal to 1"
2018-06-25 13:47:32 +0000 [error]: #0 unexpected error error_class=Aws::CloudWatchLogs::Errors::InvalidParameterException error="1 validation error detected: Value '' at 'nextToken' failed to satisfy constraint: Member must have length greater than or equal to 1"
  2018-06-25 13:47:32 +0000 [error]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/aws-sdk-core-3.11.0/lib/seahorse/client/plugins/raise_response_errors.rb:15:in `call'
  2018-06-25 13:47:32 +0000 [error]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/aws-sdk-core-3.11.0/lib/aws-sdk-core/plugins/jsonvalue_converter.rb:20:in `call'
  2018-06-25 13:47:32 +0000 [error]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/aws-sdk-core-3.11.0/lib/aws-sdk-core/plugins/idempotency_token.rb:17:in `call'
  2018-06-25 13:47:32 +0000 [error]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/aws-sdk-core-3.11.0/lib/aws-sdk-core/plugins/param_converter.rb:24:in `call'
  2018-06-25 13:47:32 +0000 [error]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/aws-sdk-core-3.11.0/lib/aws-sdk-core/plugins/response_paging.rb:10:in `call'
  2018-06-25 13:47:32 +0000 [error]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/aws-sdk-core-3.11.0/lib/seahorse/client/plugins/response_target.rb:23:in `call'
  2018-06-25 13:47:32 +0000 [error]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/aws-sdk-core-3.11.0/lib/seahorse/client/request.rb:70:in `send_request'
  2018-06-25 13:47:32 +0000 [error]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/aws-sdk-cloudwatchlogs-1.3.0/lib/aws-sdk-cloudwatchlogs/client.rb:1154:in `get_log_events'
  2018-06-25 13:47:32 +0000 [error]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluent-plugin-cloudwatch-logs-0.5.0/lib/fluent/plugin/in_cloudwatch_logs.rb:146:in `get_events'
  2018-06-25 13:47:32 +0000 [error]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluent-plugin-cloudwatch-logs-0.5.0/lib/fluent/plugin/in_cloudwatch_logs.rb:112:in `block in run'
  2018-06-25 13:47:32 +0000 [error]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluent-plugin-cloudwatch-logs-0.5.0/lib/fluent/plugin/in_cloudwatch_logs.rb:110:in `each'
  2018-06-25 13:47:32 +0000 [error]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluent-plugin-cloudwatch-logs-0.5.0/lib/fluent/plugin/in_cloudwatch_logs.rb:110:in `run'
  2018-06-25 13:47:32 +0000 [error]: #0 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.0.2/lib/fluent/plugin_helper/thread.rb:78:in `block in thread_create'
2018-06-25 13:47:32 +0000 [error]: #0 unexpected error error_class=Aws::CloudWatchLogs::Errors::InvalidParameterException error="1 validation error detected: Value '' at 'nextToken' failed to satisfy constraint: Member must have length greater than or equal to 1"
``` 

The configuration is very simple:
```
<source>
  @type cloudwatch_logs
  region eu-west-1
  tag devops.cloudwatch_logs.tst.systems
  log_group_name logs-systems
  log_stream_name logs
  use_log_stream_name_prefix true
  state_file /tmp/fluent-plugin-cloudwatch-tst-systems.state
  format none
</source>
```

If I remove the state_file, it will fetch again all the logs but after any restart of the fluent service, it doesn't start again. Any ideas on how we can further debug this?

**Comment by nunofernandes** [comment]
Here is a trace of the HTTP connection to AWS cloudwatch servers (without credentials):

```
POST / HTTP/1.1
Content-Type: application/x-amz-json-1.1
Accept-Encoding:
User-Agent: aws-sdk-ruby3/3.11.0 ruby/2.4.2 x86_64-linux aws-sdk-cloudwatchlogs/1.3.0
X-Amz-Target: Logs_20140328.GetLogEvents
Host: logs.eu-west-1.amazonaws.com
X-Amz-Date: 20180625T193102Z
X-Amz-Security-Token: FQoDXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX=
X-Amz-Content-Sha256: dba7606XXXXXXXXXXXXXXXXXXXXXXXXf33d89711af1b7a39ee14568
Authorization: AWS4-HMAC-SHA256 Credential=ASIXXXXXXXXNA/20180625/eu-west-1/logs/aws4_request, SignedHeaders=content-type;host;x-amz-content-sha256;x-amz-date;x-amz-security-token;x-amz-target, Signature=7ce86cXXXXXXXXXXXXXXXXXXXX446de3562ce40fd9f
Content-Length: 125
Accept: */*

{"logGroupName":"logs-systems","logStreamName":"logs-docs-site-dev/docs/0abdc2fd-b82e-41cb-b7ef-d7d5ffde5348","nextToken":""}HTTP/1.1 400 Bad Request
x-amzn-RequestId: 4bbd9bc6-78ae-11e8-a3f1-d39b3c4dbb70
Content-Type: application/x-amz-json-1.1
Content-Length: 184
Date: Mon, 25 Jun 2018 19:31:02 GMT
Connection: close

{"__type":"InvalidParameterException","message":"1 validation error detected: Value '' at 'nextToken' failed to satisfy constraint: Member must have length greater than or equal to 1"}
```

**Comment by barrytam20** [comment]
i'm observing this behavior on `'fluent-plugin-cloudwatch-logs' version '0.6.0'` as well

**Comment by sergeyarl** [comment]
same for me

**Comment by barrytam20** [comment]
for me, it works for a while then this error appears. Perhaps `store_next_token` is writing an empty string into the state file?

**Comment by cosmo0920** [comment]
Fixed in #122.
